src/parser/extensions.rs

```diff
@@ -9,40 +9,53 @@ use xml::reader::XmlEvent;
 use crate::errors::{GpxError, GpxResult};
 use crate::parser::Context;
 
+use super::verify_starting_tag;
+
 /// consume consumes a single string as tag content.
 pub fn consume<R: Read>(context: &mut Context<R>) -> GpxResult<()> {
-    let mut started = false;
+    verify_starting_tag(context, "extensions")?;
+
+    // helper to keep track of inner tags until generic parser exists for extensions-content
+    let mut inner_tag_stack = Vec::<String>::new();
 
-    for event in context.reader() {
+    for event in &mut context.reader {
         match event? {
             XmlEvent::StartElement { name, .. } => {
-                // flip started depending on conditions
-                if &name.local_name == "extensions" {
-                    if started {
-                        return Err(GpxError::TagOpenedTwice("extensions"));
+                // push every opening-element on the stack
+                // we treat inner "extensions"-tags as any other tag
+                let child = name.local_name.clone();
+                inner_tag_stack.push(child);
+            },
+            XmlEvent::EndElement { name } => {
+                // as long as there is an inner tag open..
+                if let Some(current_inner_tag) = inner_tag_stack.pop() {
+                    // the closing tag has to match the current open tag, this makes it also impossible to close the "extensions" while an inner tag is open
```

Comment:
If the XML library doesn't allow a mismatched closing tag, I think we shouldn't handle it here. Then the stack isn't necessary.

Instead I was thinking of returning from here when the depth reaches 0 (meaning that the outer `<extensions>` was closed. I don't know when the `for` loop finishes. Can there be something else after the final `</extensions>`?

If you look closely, the old code does a `return Ok(());`. Your first version did not.

Reply:
> If the XML library doesn't allow a mismatched closing tag, I think we shouldn't handle it here. Then the stack isn't necessary.

For code-clarity this should be handled here I guess, it should not cost much, hence the will not be huge extensions-tags. But if you want I can remove it of course.

> Instead I was thinking of returning from here when the depth reaches 0 (meaning that the outer <extensions> was closed. I don't know when the for loop finishes. Can there be something else after the final </extensions>?

I think the approach with counting the depth was not the best, hence it handles internal "extensions"-tags in a special way, different from the other internal tags.
I guess, the code should handle inner extensions-tags like any other inner tag, hence they depend on the extensions-schema, not the tag-name. The new code is much more generic and aligned with the coding-style of the other tag-parsers.
Something after the final </extensions> is (as a consequence of xml-schema and the visitor-pattern) always outside of the extensions-parser. This would have been the same with the depth approach.
This is then handled by the "outer" parser, in this case the track-parser and so on..

> If you look closely, the old code does a return Ok(());. Your first version did not.

Thats true, I got your point with returning ```Ok(())``` where you suggested it in the old code and I agree, this would have been correct.

Reply:
Okay, so I tried to add a mismatched closing tag to a test, and got:

```
[src/parser/extensions.rs:77] &result = Err(
    XmlParseError(
        Error {
            pos: 5:22,
            kind: Syntax(
                "Unexpected closing tag: foo, expected extensions",
            ),
        },
    ),
)
```

And the `if let Some(current_inner_tag) = inner_tag_stack.pop()` check never sees the unmatched `</foo>. So `xml-rs` already handles this case, and I don't think we should do it a second time.

I also skimmed the rest of the code, and didn't find another place which builds a tag stack like this.

So if I'm not mistaken, that's dead code, and I think we should drop it. It's not just about the performance, but it's also misleading for the readers of the code.

My suggestion was to do something like:

```rust
/// consume consumes a single string as tag content.
pub fn consume<R: Read>(context: &mut Context<R>) -> GpxResult<()> {
    verify_starting_tag(context, "extensions")?;

    let mut depth = 1;
    for event in &mut context.reader {
        match event? {
            XmlEvent::StartElement { name, .. } => {
                // keep track of how many `extension` tags we are under
                if name.local_name == "extensions" {
                    depth += 1;
                }
            }
            XmlEvent::EndElement { name } => {
                dbg!(&name);
                if name.local_name == "extensions" {
                    // pop one
                    depth -= 1;
                }
                if depth == 0 {
                    return Ok(());
                }
            }
            _ => {}
        }
    }

    Err(GpxError::MissingClosingTag("extensions"))
}
```

Is this incorrect in some cases?

Reply:
> So if I'm not mistaken, that's dead code, and I think we should drop it. It's not just about the performance, but it's also misleading for the readers of the code.

Yes ok, I will remove it.


> Is this incorrect in some cases? Or is there any case when my version and yours produce different results?:

No, I do not think it is incorrect in an algorithmic way. It would work fine I guess.

I was just thinking, that all tags inside the extensions-tag should be treated the same way by the code. In the future the code should parse all the inner tags into a generic structure, that represents the all the tags inside the "extensions" to have it accessible for the user.
So in the future - if we migh have some generic-parser - we can, instead of 
```rust
            XmlEvent::StartElement { name, .. } => {
                // push every opening-element on the stack
                // we treat inner "extensions"-tags as any other tag
                let child = name.local_name.clone();
                inner_tag_stack.push(child);
            },
```

just write
```rust
            XmlEvent::StartElement { name, .. } => {
                // handle attributes etc ... 
                // ..
                let child = name.local_name.clone();
                generic_whatever::consume(child)
            },
```

and drop the whole inner stack/depth-tracking completely.


```rust
/// consume consumes a single string as tag content.
pub fn consume<R: Read>(context: &mut Context<R>) -> GpxResult<()> {
    verify_starting_tag(context, "extensions")?;

    let mut depth = 1;
    for event in &mut context.reader {
        match event? {
            XmlEvent::StartElement { name, .. } => {
                // I think its bad to hardcode the check on name == "extensions", because it is not a generic approach
                // and treats inner tags that are called "extensions" differently from any other inner tags, like "a", "foo", "bar"
                // It is correct, but feels wrong, maybe only a personal feeling
                if name.local_name == "extensions" {
                    depth += 1;
                }
            }
            XmlEvent::EndElement { name } => {
                if name.local_name == "extensions" {
                    // pop one
                    depth -= 1;
                    if depth == 0 {
                        return Ok(());
                    }
                }
            }
            _ => {}
        }
    }

    Err(GpxError::MissingClosingTag("extensions"))
}
```

I see the inner_tag vector more as a placeholder for an inner parser call. It gives a direction how I personally intended to go on with handling extensions inner content in another pullrequest based on this one.

In my opinion, depth tracking points in another direction, where the extensions-parser has to "know" something about the extensions content.

Reply:
I'm not sure how that's supposed to work. This `consume` starts with an `<extensions>` and needs to parse (well, drop) everything until the matching `</extensions>`. We need to track the depth to achieve that (the parser won't do it for  us), regardless of the extension schema.

If we want to expose the contents of those to the user, agreed, that's not enough, but the stack is not helping with that, and it doesn't buy us _anything_, since the check you've added already exists in the parser.

Reply:
I mean, the "ideal" API would be something like:

```rust
trait ExtensionParser {
    fn consume<R: Read>(&mut self, context: &mut Context<R>) -> GpxResult<()>;
}

struct Extensions;

impl ExtensionParser for Extensions {
    // ...
}

    let extension_parser = Extensions;
    let gpx: Gpx = GpxReader::new().with_extension_parser(&mut extension_parser).read(reader).unwrap();
```

, right?

And we don't need that stack to implement it.